File: src/main/java/com/hoi4utils/clausewitz/map/province/Province.java
package com.hoi4utils.clausewitz.map.province;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.function.Consumer;

public class Province implements Iterable<Province> {
	static final List<Province> provinces = new ArrayList<>();
	private final int _id;

	protected Province(int provinceid) throws IllegalArgumentException {
		if (provinceid < 0) {
			throw new IllegalArgumentException("Province id must be non-negative");
		}
		this._id = provinceid;
	}

	public static Province of(int provinceid) throws IllegalArgumentException {
		for (Province province : provinces) {
			if (province._id == provinceid) {
				return province;
			}
		}
		return new Province(provinceid);
	}

	@Override
	public @NotNull Iterator<Province> iterator() {
		return provinces.iterator();
	}

	@Override
	public void forEach(Consumer<? super Province> action) {
		provinces.forEach(action);
	}

	@Override
	public Spliterator<Province> spliterator() {
		return provinces.spliterator();
	}

	public static List<Province> list() {
		return new ArrayList<>(provinces);
	}

	public int id() {
		return _id;
	}

	public String idStr() { return Integer.toString(_id); }
}

File: src/main/scala/com/hoi4utils/clausewitz/map/province/Province.scala
package com.hoi4utils.clausewitz.map.province

import org.jetbrains.annotations.NotNull
import org.apache.logging.log4j.{LogManager, Logger}

import scala.collection.mutable.ListBuffer

class Province private(private val _id: Int) extends Comparable[Province] {
  /* init */
  require(_id >= 0, "Province id must be non-negative")

  Province.add(this)

  def id = _id

  override def compareTo(o: Province): Int = this.id.compareTo(o.id)
}

object Province {
  val LOGGER: Logger = LogManager.getLogger(classOf[Province])
  private val provinces = new ListBuffer[Province]()

  /**
   * Clears all provinces and any other relevant values.
   */
  def clear(): Unit = {
    provinces.clear()
  }

  def add(province: Province): List[Province] = {
    provinces += province
    provinces.toList
  }

  def list: List[Province] = provinces.toList

  def apply(id: Int): Province = {
    require(id >= 0, "Province id must be non-negative")
    provinces.find(_.id == id) match {
      case Some(existing) =>
        // Return the already-created instance
        existing
      case None =>
        // Create, cache, and return a new Province
        val newProvince = new Province(id)
        provinces += newProvince
        newProvince
    }
  }
}
